### ISSUE TITLE
"Duplicate value: http" error when it is declared in service.yml file

### ISSUE THREAD

**Issue by rmarting** [eclipse/jkube#450]
### Description
Declared a ```deployment.yml``` file with the following structure:

```
spec:
  template:
    spec:
      containers:
        - ports:
          - containerPort: 8181
            name: http
            protocol: TCP
```

Throws the following exception when the application is deployed:

```
[ERROR] oc: Failed to update DeploymentConfig from openshift.yml. io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PUT at: https://api.crc.testing:6443/apis/apps.openshift.io/v1/namespaces/amq-streams-demo/deploymentconfigs/kafka-clients-quarkus-sample-green. Message: DeploymentConfig.apps.openshift.io "kafka-clients-quarkus-sample-green" is invalid: spec.template.spec.containers[0].ports[1].name: Duplicate value: "http". Received status: Status(apiVersion=v1, code=422, details=StatusDetails(causes=[StatusCause(field=spec.template.spec.containers[0].ports[1].name, message=Duplicate value: "http", reason=FieldValueDuplicate, additionalProperties={})], group=apps.openshift.io, kind=DeploymentConfig, name=kafka-clients-quarkus-sample-green, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=DeploymentConfig.apps.openshift.io "kafka-clients-quarkus-sample-green" is invalid: spec.template.spec.containers[0].ports[1].name: Duplicate value: "http", metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Invalid, status=Failure, additionalProperties={}).. DeploymentConfig(apiVersion=apps.openshift.io/v1, kind=DeploymentConfig, metadata=ObjectMeta(annotations={app.openshift.io/vcs-ref=feature/b-g-deployment-strategy, jkube.io/git-url=https://github.com/rmarting/kafka-clients-quarkus-sample.git, app.openshift.io/vcs-uri=https://github.com/rmarting/kafka-clients-quarkus-sample.git, jkube.io/git-commit=c2c0cb82670db43882c8b3b36e431173a0501927, jkube.io/git-branch=feature/b-g-deployment-strategy}, clusterName=null, creationTimestamp=null, deletionGracePeriodSeconds=null, deletionTimestamp=null, finalizers=[], generateName=null, generation=null, labels={group=com.rmarting.kafka, project=kafka-clients-quarkus-sample, version=1.0.0-SNAPSHOT, app=kafka-clients-quarkus-sample-green, provider=jkube}, managedFields=[], name=kafka-clients-quarkus-sample-green, namespace=null, ownerReferences=[], resourceVersion=null, selfLink=null, uid=null, additionalProperties={}), spec=DeploymentConfigSpec(minReadySeconds=null, paused=null, replicas=1, revisionHistoryLimit=2, selector={app=kafka-clients-quarkus-sample, provider=jkube, group=com.rmarting.kafka}, strategy=DeploymentStrategy(activeDeadlineSeconds=null, annotations=null, customParams=null, labels=null, recreateParams=null, resources=null, rollingParams=RollingDeploymentStrategyParams(intervalSeconds=null, maxSurge=null, maxUnavailable=null, post=null, pre=null, timeoutSeconds=3600, updatePeriodSeconds=null, additionalProperties={}), type=Rolling, additionalProperties={}), template=PodTemplateSpec(metadata=ObjectMeta(annotations={app.openshift.io/vcs-ref=feature/b-g-deployment-strategy, jkube.io/git-url=https://github.com/rmarting/kafka-clients-quarkus-sample.git, app.openshift.io/vcs-uri=https://github.com/rmarting/kafka-clients-quarkus-sample.git, jkube.io/git-commit=c2c0cb82670db43882c8b3b36e431173a0501927, jkube.io/git-branch=feature/b-g-deployment-strategy}, clusterName=null, creationTimestamp=null, deletionGracePeriodSeconds=null, deletionTimestamp=null, finalizers=[], generateName=null, generation=null, labels={app=kafka-clients-quarkus-sample, provider=jkube, version=1.0.0-SNAPSHOT, group=com.rmarting.kafka}, managedFields=[], name=null, namespace=null, ownerReferences=[], resourceVersion=null, selfLink=null, uid=null, additionalProperties={}), spec=PodSpec(activeDeadlineSeconds=null, affinity=null, automountServiceAccountToken=null, containers=[Container(args=[], command=[], env=[EnvVar(name=KAFKA_USER_PASSWORD, value=null, valueFrom=EnvVarSource(configMapKeyRef=null, fieldRef=null, resourceFieldRef=null, secretKeyRef=SecretKeySelector(key=password, name=application, optional=null, additionalProperties={}), additionalProperties={}), additionalProperties={}), EnvVar(name=APP_BG_VERSION, value=green, valueFrom=null, additionalProperties={}), EnvVar(name=APP_BG_MODE, value=offline, valueFrom=null, additionalProperties={}), EnvVar(name=KUBERNETES_NAMESPACE, value=null, valueFrom=EnvVarSource(configMapKeyRef=null, fieldRef=ObjectFieldSelector(apiVersion=null, fieldPath=metadata.namespace, additionalProperties={}), resourceFieldRef=null, secretKeyRef=null, additionalProperties={}), additionalProperties={})], envFrom=[], image=kafka-clients-quarkus-sample-green:1.0.0-SNAPSHOT, imagePullPolicy=IfNotPresent, lifecycle=null, livenessProbe=Probe(exec=null, failureThreshold=3, httpGet=HTTPGetAction(host=null, httpHeaders=[], path=/health/live, port=IntOrString(IntVal=8181, Kind=0, StrVal=null, additionalProperties={}), scheme=HTTP, additionalProperties={}), initialDelaySeconds=5, periodSeconds=10, successThreshold=1, tcpSocket=null, timeoutSeconds=1, additionalProperties={}), name=quarkus, ports=[ContainerPort(containerPort=8181, hostIP=null, hostPort=null, name=http, protocol=TCP, additionalProperties={}), ContainerPort(containerPort=8080, hostIP=null, hostPort=null, name=http, protocol=TCP, additionalProperties={})], readinessProbe=Probe(exec=null, failureThreshold=3, httpGet=HTTPGetAction(host=null, httpHeaders=[], path=/health/ready, port=IntOrString(IntVal=8181, Kind=0, StrVal=null, additionalProperties={}), scheme=HTTP, additionalProperties={}), initialDelaySeconds=5, periodSeconds=10, successThreshold=1, tcpSocket=null, timeoutSeconds=1, additionalProperties={}), resources=null, securityContext=SecurityContext(allowPrivilegeEscalation=null, capabilities=null, privileged=false, procMount=null, readOnlyRootFilesystem=null, runAsGroup=null, runAsNonRoot=null, runAsUser=null, seLinuxOptions=null, windowsOptions=null, additionalProperties={}), startupProbe=null, stdin=null, stdinOnce=null, terminationMessagePath=null, terminationMessagePolicy=null, tty=null, volumeDevices=[], volumeMounts=[], workingDir=null, additionalProperties={})], dnsConfig=null, dnsPolicy=null, enableServiceLinks=null, ephemeralContainers=[], hostAliases=[], hostIPC=null, hostNetwork=null, hostPID=null, hostname=null, imagePullSecrets=[], initContainers=[], nodeName=null, nodeSelector=null, overhead=null, preemptionPolicy=null, priority=null, priorityClassName=null, readinessGates=[], restartPolicy=null, runtimeClassName=null, schedulerName=null, securityContext=null, serviceAccount=null, serviceAccountName=null, shareProcessNamespace=null, subdomain=null, terminationGracePeriodSeconds=null, tolerations=[], topologySpreadConstraints=[], volumes=[], additionalProperties={}), additionalProperties={}), test=null, triggers=[DeploymentTriggerPolicy(imageChangeParams=null, type=ConfigChange, additionalProperties={}), DeploymentTriggerPolicy(imageChangeParams=DeploymentTriggerImageChangeParams(automatic=true, containerNames=[quarkus], from=ObjectReference(apiVersion=null, fieldPath=null, kind=ImageStreamTag, name=kafka-clients-quarkus-sample-green:1.0.0-SNAPSHOT, namespace=null, resourceVersion=null, uid=null, additionalProperties={}), lastTriggeredImage=null, additionalProperties={}), type=ImageChange, additionalProperties={})], additionalProperties={}), status=null, additionalProperties={}): io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PUT at: https://api.crc.testing:6443/apis/apps.openshift.io/v1/namespaces/amq-streams-demo/deploymentconfigs/kafka-clients-quarkus-sample-green. Message: DeploymentConfig.apps.openshift.io "kafka-clients-quarkus-sample-green" is invalid: spec.template.spec.containers[0].ports[1].name: Duplicate value: "http". Received status: Status(apiVersion=v1, code=422, details=StatusDetails(causes=[StatusCause(field=spec.template.spec.containers[0].ports[1].name, message=Duplicate value: "http", reason=FieldValueDuplicate, additionalProperties={})], group=apps.openshift.io, kind=DeploymentConfig, name=kafka-clients-quarkus-sample-green, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=DeploymentConfig.apps.openshift.io "kafka-clients-quarkus-sample-green" is invalid: spec.template.spec.containers[0].ports[1].name: Duplicate value: "http", metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Invalid, status=Failure, additionalProperties={}).
```

The next definition works successfully: 

```
spec:
  template:
    spec:
      containers:
        - ports:
          - containerPort: 8181
            name: http-web
            protocol: TCP
```

However the deployment declares finally two ports (http:8080) and (http-web:8181), but my applications has not declared that port as needed.

```
spec:
  template:
    spec:
      containers:
        - ports:
            - name: http-web
              containerPort: 8181
              protocol: TCP
            - name: http
              containerPort: 8080
              protocol: TCP
```

IMHO my deployment definition should be used to define the service ports.

### Info

* Eclipse JKube version : openshift-maven-plugin:1.0.1
* Maven version (`mvn -v`) :
```
❯ mvn -v
Maven home: /home/rmarting/Software/apache-maven-3.6.3
Java version: 11.0.8, vendor: Red Hat, Inc., runtime: /usr/lib/jvm/java-11-openjdk-11.0.8.10-2.fc32.x86_64
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.8.15-201.fc32.x86_64", arch: "amd64", family: "unix"
```
* Kubernetes / Red Hat OpenShift setup and version :
```
❯ crc version
CodeReady Containers version: 1.17.0+99f5c87
OpenShift version: 4.5.14 (embedded in binary)
```
* If it's a bug, how to reproduce :

1.- Create a deployment.yml file with a http port definition
2.- Deploy application

**Comment by manusa** [comment]
Port should be configured at image level. Then both the generated Deployment and Service configurations will pick up the value.

So in your case I guess setting this property or XML configuration `jkube.generator.quarkus.webPort=8181` should work without the need of that extra fragment.

Quarkus integration is still in early stage. This value should be inferred directly from the [application.properties](https://quarkus.io/guides/all-config#quarkus-vertx-http_quarkus.http.port).

**Comment by rmarting** [comment]
Adding in ```pom.xml``` file the property ```<jkube.generator.quarkus.webPort>8181</jkube.generator.quarkus.webPort>``` and removing the ```ports``` definition in ```deployment.yml``` file the application is deployed successfully and only one port definition is created in the deployment object.

However removing the ```ports``` definition in the ```service.yml``` file I got the following exception:

```
[ERROR] Failed to execute goal org.eclipse.jkube:openshift-maven-plugin:1.0.1:apply (default-cli) on project kafka-clients-quarkus-sample: Failed to update Service from openshift.yml. io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PATCH at: https://api.crc.testing:6443/api/v1/namespaces/amq-streams-demo/services/kafka-clients-quarkus-sample-green. Message: Service "kafka-clients-quarkus-sample-green" is invalid: spec.ports: Required value. Received status: Status(apiVersion=v1, code=422, details=StatusDetails(causes=[StatusCause(field=spec.ports, message=Required value, reason=FieldValueRequired, additionalProperties={})], group=null, kind=Service, name=kafka-clients-quarkus-sample-green, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=Service "kafka-clients-quarkus-sample-green" is invalid: spec.ports: Required value, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Invalid, status=Failure, additionalProperties={}).. Service(apiVersion=v1, kind=Service, metadata=ObjectMeta(annotations={app.openshift.io/vcs-ref=feature/b-g-deployment-strategy, jkube.io/git-url=https://github.com/rmarting/kafka-clients-quarkus-sample.git, app.openshift.io/vcs-uri=https://github.com/rmarting/kafka-clients-quarkus-sample.git, jkube.io/git-commit=c2c0cb82670db43882c8b3b36e431173a0501927, jkube.io/git-branch=feature/b-g-deployment-strategy}, clusterName=null, creationTimestamp=null, deletionGracePeriodSeconds=null, deletionTimestamp=null, finalizers=[], generateName=null, generation=null, labels={app=kafka-clients-quarkus-sample-green, group=com.rmarting.kafka, project=kafka-clients-quarkus-sample-green, provider=jkube, expose=true, version=1.0.0-SNAPSHOT}, managedFields=[], name=kafka-clients-quarkus-sample-green, namespace=null, ownerReferences=[], resourceVersion=624677, selfLink=null, uid=null, additionalProperties={}), spec=ServiceSpec(clusterIP=null, externalIPs=[], externalName=null, externalTrafficPolicy=null, healthCheckNodePort=null, ipFamily=null, loadBalancerIP=null, loadBalancerSourceRanges=[], ports=[], publishNotReadyAddresses=null, selector={deploymentconfig=kafka-clients-quarkus-sample-green, app=kafka-clients-quarkus-sample, provider=jkube, group=com.rmarting.kafka}, sessionAffinity=null, sessionAffinityConfig=null, topologyKeys=[], type=NodePort, additionalProperties={}), status=null, additionalProperties={}) -> [Help 1]
```

If I understood well, with the ```jkube.generator.quarkus.webPort``` property should be enought for both objects, right?

**Comment by manusa** [comment]
> However removing the ports definition in the service.yml file I got the following exception:...

I think this may be related to the previous (#448) issue.

The information for all of the enrichers is read from the applicable ImageConfiguration. So probably there's an issue when merging your fragment.

I'll try to reproduce on Monday.

**Comment by manusa** [comment]
After #448 fix, I cannot reproduce this issue anymore.

Providing a `service.yml` fragment as follows:
```yml
metadata:
  name: ${project.artifactId}
  labels:
    app: ${project.artifactId}
    group: ${project.groupId}
    project: ${project.artifactId}
    provider: jkube
    expose: "true"
spec:
  type: ${jkube.enricher.jkube-service.type}
  selector:
    deploymentconfig: ${project.artifactId}
```

With the mentioned properties in my `pom.xml`:
```xml
<jkube.generator.quarkus.webPort>8181</jkube.generator.quarkus.webPort>
<jkube.enricher.jkube-service.type>NodePort</jkube.enricher.jkube-service.type>
```

Generates the following `service.yml`:
```yml
---
apiVersion: v1
kind: Service
metadata:
  annotations:
    jkube.io/git-url: git@github.com:manusa/jkube.git
    jkube.io/git-commit: 6f926aa4e8d1ceba270102fdcef64d52050487bc
    jkube.io/git-branch: fix/watch-rolling-update
  labels:
    app: quarkus
    group: org.eclipse.jkube.quickstarts.maven
    project: quarkus
    provider: jkube
    expose: "true"
    version: 1.0.2-SNAPSHOT
  name: quarkus
spec:
  ports:
  - name: intermapper
    port: 8181
    protocol: TCP
    targetPort: 8181
  selector:
    deploymentconfig: quarkus
    app: quarkus
    provider: jkube
    group: org.eclipse.jkube.quickstarts.maven
  type: NodePort
```

However, I tried the same with 1.0.1, but can't reproduce the reported misbehavior either.

**Comment by rmarting** [comment]
Did you tried with the following ```service.yml``` fragment?

```
spec:
  template:
    spec:
      containers:
        - ports:
          - containerPort: 8181
            name: http
            protocol: TCP
```

Here I am defining a port named as ```http```, which throws the issue described.

**Comment by manusa** [comment]
Sorry. In my [last comment](https://github.com/eclipse/jkube/issues/450#issuecomment-717282200) I was referring about the issue when providing a `service.yml` with no port definitions. The final service is correctly populated with the ports inferred in the image configuration.

> Did you tried with the following service.yml fragment?...

I think this fragment belongs to a deployment, not a service (typo?).

If you intended to mean `deployment.yml` instead, yes I did. But as I stated in the [first reply](https://github.com/eclipse/jkube/issues/450#issuecomment-715321676), port configuration should be performed at Image Configuration level (via properties, xml, etc.). If we were to infer port configuration from fragments, we would have to process them several times, and we'd face many configuration collisions which would not be that straightforward to solve.

IMO the only issue at hand is what was reported [here](https://github.com/eclipse/jkube/issues/450#issuecomment-715381318), but I'm not able to reproduce this. On the other hand, if you need to redefine the port name, there is already another issue for this #367 (out of curiosity, why do you need to rename the ports?)

**Comment by rmarting** [comment]
Sorry @manusa ... I read and replied so fast without review deeply your comments. The issue is reproducible with the following set up:

* Deploying into OpenShift
* Not declared the ```kube.generator.quarkus.webPort``` property in the ```pom.xml``` file
* Using this ```deployment.yml``` fragment:

```
metadata:
  labels:
    group: ${project.groupId}
    project: ${project.artifactId}
    version: ${project.version}
    provider: jkube
spec:
  triggers:
    - type: ConfigChange
  template:
    spec:
      containers:
        - env:
          - name: KAFKA_USER_PASSWORD
            valueFrom:
              secretKeyRef:
                key: password
                name: application
          ports:
          - containerPort: 8181
            name: http
            protocol: TCP
          livenessProbe:
            failureThreshold: 3
            httpGet:
              path: /health/live
              port: 8181
              scheme: HTTP
            initialDelaySeconds: 5
            periodSeconds: 10
            successThreshold: 1
            timeoutSeconds: 1
          readinessProbe:
            failureThreshold: 3
            httpGet:
              path: /health/ready
              port: 8181
              scheme: HTTP
            initialDelaySeconds: 5
            periodSeconds: 10
            successThreshold: 1
            timeoutSeconds: 1
```

* Using this ```service.yml``` fragment:

```
metadata:
  labels:
    group: ${project.groupId}
    project: ${project.artifactId}
    provider: jkube
    expose: "true"
spec:
  ports:
    - name: http
      port: 8181
      protocol: TCP
      targetPort: 8181
  selector:
    deploymentconfig: ${project.artifactId}
```

The log message is:

```
[ERROR] oc: Failed to update DeploymentConfig from openshift.yml. io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PUT at: https://api.crc.testing:6443/apis/apps.openshift.io/v1/namespaces/amq-streams-demo/deploymentconfigs/kafka-clients-quarkus-sample-green. Message: DeploymentConfig.apps.openshift.io "kafka-clients-quarkus-sample-green" is invalid: spec.template.spec.containers[0].ports[1].name: Duplicate value: "http".
```

I was not aware about the issue #367 but this is my use:

I want to declare other web port, instead of 8080, because I am not using it. Basically I would like to redefine the default web port (not rename the original http port). My deployment declares the http port, however jkube declares always the default http port (8080) whatever other ports I am adding in my deployment descriptor. So Could I redefine the default http port in my deployment without use ```jkube.generator.quarkus.webPort``` property?

IMHO the fragments provided by the developer should be applied on top other enrichers done by jkube.

Do you know what I mean?

**Comment by rmarting** [comment]
@manusa you are right and after reading deeper about enrichers I am understand them successfully. I come back to my case and after the following changes everything works:

* Declare in my ```pom.xml``` file the property for Quarkus:
```
    <properties>
        <jkube.generator.quarkus.webPort>8181</jkube.generator.quarkus.webPort>
    </properties>
```

* Commented the ```ports``` definition in my ```deployment.yml``` fragment file

So it seems that the Quarkus generator is not getting the right port value from my ```application.properties``` file and I have to declare it in the ```pom.xml``` file. So I guess the root cause of this issue is there. Am I wrong?

**Comment by manusa** [comment]
You're right. We definitely need to improve the Quarkus Generator.
Thx for taking the time to look into this.

**Comment by manusa** [comment]
fixed by #623
